File: spotcheck-ai/backend/app/core/geo.py
"""座標計算。PostGISは導入せず、バウンディングボックス＋Haversineで扱う（docs/02-database.md 2.2）。"""

from __future__ import annotations

import math

EARTH_RADIUS_M = 6_371_000.0
#: 緯度1度あたりの距離（約111.32km）
METERS_PER_DEGREE_LAT = 111_320.0


def haversine_meters(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """2点間の大円距離をメートルで返す。"""
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    d_phi = phi2 - phi1
    d_lambda = math.radians(lng2 - lng1)
    a = math.sin(d_phi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(d_lambda / 2) ** 2
    return 2 * EARTH_RADIUS_M * math.asin(math.sqrt(a))


def bounding_box(lat: float, lng: float, radius_km: float) -> tuple[float, float, float, float]:
    """粗い絞り込み用の矩形 (min_lat, max_lat, min_lng, max_lng) を返す。

    経度方向の1度あたりの距離は緯度によって変わるため cos で補正する。
    高緯度・極付近では補正値が0に近づくため、下限を設けて全経度を含める。
    """
    radius_m = radius_km * 1000
    d_lat = radius_m / METERS_PER_DEGREE_LAT
    cos_lat = math.cos(math.radians(lat))
    if abs(cos_lat) < 1e-6:
        return (max(-90.0, lat - d_lat), min(90.0, lat + d_lat), -180.0, 180.0)
    d_lng = radius_m / (METERS_PER_DEGREE_LAT * abs(cos_lat))
    return (
        max(-90.0, lat - d_lat),
        min(90.0, lat + d_lat),
        max(-180.0, lng - d_lng),
        min(180.0, lng + d_lng),
    )

File: spotcheck-ai/frontend/src/lib/geo.ts
/** 距離計算・座標フォーマット。 */

const EARTH_RADIUS_M = 6_371_000;
/** 徒歩の想定速度。画面⑤の所要時間の概算に使う（docs/05-frontend.md 画面⑤）。 */
export const WALKING_METERS_PER_MINUTE = 80;

export function haversineMeters(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dPhi = toRad(lat2 - lat1);
  const dLambda = toRad(lng2 - lng1);
  const a =
    Math.sin(dPhi / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLambda / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.sqrt(a));
}

export function formatDistance(km: number): string {
  return km < 1 ? `${Math.round(km * 1000)} m` : `${km.toFixed(1)} km`;
}

/** 「1.2 km（徒歩約15分）」の形式。徒歩80m/分で概算した値であることを示す。 */
export function formatDistanceWithWalk(km: number): string {
  const minutes = Math.max(1, Math.round((km * 1000) / WALKING_METERS_PER_MINUTE));
  return `${formatDistance(km)}（徒歩約${minutes}分）`;
}

export function formatCoords(lat: number, lng: number): string {
  return `${lat.toFixed(5)}, ${lng.toFixed(5)}`;
}
